src/starks/constraints/boundary.rs

```diff
@@ -59,6 +60,22 @@ impl<F: IsField> BoundaryConstraints<F> {
             .collect()
     }
 
+    pub fn steps_for_boundary(&self) -> Vec<usize> {
+        self.constraints
+            .iter()
+            .unique_by(|elem| elem.step)
+            .map(|v| v.step)
+            .collect()
+    }
```

Comment:
There are two uses of this and both only iterate once over the result. Consider returning an iterator instead:
```suggestion
    pub fn steps_for_boundary(&self) -> impl Iterator<Item = usize> {
        self.constraints
            .iter()
            .cloned()
            .unique_by(|elem| elem.step)
            .map(|v| v.step)
    }
```